src/bms/command/channel/converter.rs

```diff
@@ -1,118 +1,74 @@
-//! For one-way converting key/channel, please see [`KeyLayoutConverter`] trait.
+//! For one-way converting key/channel, please see [`KeyMappingConverter`] trait.
 
 use std::collections::HashMap;
 
 use super::{Key, PlayerSide, mapper::KeyMapping};
+use crate::bms::ast::rng::JavaRandom;
+use crate::bms::command::time::ObjTime;
 
-/// A trait for converting [`KeyMapping`]s.
+/// A trait for converting [`KeyMapping`]s in different layouts.
 ///
-/// - Difference from [`KeyLayoutMapper`]:
-///   - [`KeyLayoutMapper`] can convert between different key channel modes. It's two-way.
-///   - [`KeyLayoutConverter`] can convert into another key layout. It's one-way.
-pub trait KeyLayoutConverter {
-    /// Convert a [`KeyMapping`] to another key layout.
-    fn convert<T: KeyMapping>(&mut self, beat_map: T) -> T;
+/// - Difference from [`super::mapper::KeyLayoutMapper`]:
+///   - [`super::mapper::KeyLayoutMapper`] can convert between different key channel modes. It's two-way.
+///   - [`KeyMappingConverter`] can convert into another key layout. It's one-way.
+///   - [`KeyMappingConverter`] operates on iterators of `(ObjTime, KeyMapping)` pairs, preserving timing information.
+pub trait KeyMappingConverter {
+    /// Convert an iterator of `(ObjTime, KeyMapping)` pairs to another key layout.
+    fn convert<T: KeyMapping>(
+        &mut self,
+        mappings: impl Iterator<Item = (ObjTime, T)>,
+    ) -> impl Iterator<Item = (ObjTime, T)>;
```

Comment:
Why do `convert` require `ObjTime` as a first item of the tuple while preserving it? It should be removed like:
```suggestion
///   - [`KeyMappingConverter`] operates on iterators of `(ObjTime, KeyMapping)` pairs, preserving timing information.
pub trait KeyMappingConverter {
    /// Convert an iterator of `KeyMapping` to another key layout's one.
    fn convert<T: KeyMapping>(
        &mut self,
        mappings: impl Iterator<Item = T>,
    ) -> impl Iterator<Item = T>;
```

Reply:
From a musical point of view, the overlapped notes should be kept to play the music correctly. This would be warned by our diagnostics, but handling them (such as merging into one) are delegated to userland.
Anyway, the features to shuffle notes won't (and shouldn't) change timing of notes, I expect.

Reply:
OK. Is it suitable for adding several traits for degrees of ability? But it's a challenge for naming them.